### DISCUSSION TITLE
extendible list of parameters for the long-term

### DISCUSSION THREAD

**Discussion by yunjunz** [opera-adt/s1-reader#16] [Ideas]
Currently, we read and calculate a lot of parameters in the [`burst_from_xml`](https://github.com/opera-adt/sentinel1-reader/blob/a4bfadef4be5e750f2f696b056c8ab3d3cd19b3b/src/sentinel1_reader/sentinel1_reader.py#L250), but only part of them are passed to [`Sentinel1BurstSlc`](https://github.com/opera-adt/sentinel1-reader/blob/a4bfadef4be5e750f2f696b056c8ab3d3cd19b3b/src/sentinel1_reader/sentinel1_burst_slc.py#L164) class and available to the user. It would be more convenient if we could expose more, or maybe all of them, to the user. 

Related, we would also expect the list of parameters to grow and be grouped in the long term, as the users might want more parameters, and the data providers, e.g. ESA, may provide more annotation data.

The long list of parameters between `burst_from_xml()` and `Sentienl1BurstSlc` could be error-prone and not friendly to maintain in the long term. How about moving the majority of XML parsing into the burst SLC class, as outlined below?

+ `burst_from_xml()`:
    - use the main body to extract a minimum list of parameters, including the list of burst_id
    - then pass the annotation_path, orbit_path, tiff_path, and burst_id to the burst SLC class, for initiation. Then run something like `slc_burst.populate_metadata()` for the actual extraction.
+ `Sentinel1BurstSlc`: 
    - add a member function, e.g. `populate_metadata()`, to extract the parameters from the annotation files, instead of grabbing them as inputs. In this way, we only need to modify here to add new fields, which will be readily accessible to users.
    - we could add similar function, e.g. `populate_rfi_metadata()`, for the future annotation file.

This is just the basic idea, the details are surely to be determined. What do you think @opera-adt/opera-adt ?

**Comment by LiangJYu** [comment]
I think the following half-baked ideas could be useful:
1. Adding a `xml_to_burst` `@classmethod` in `Sentinel1BurstSlc`. e.g.
```python
@dataclass
class BurstSlc:
    id: str

    @classmethod
    def burst_from_file(cls, zip_path):
        # do ZIP/XML parsing
        return cls(id)

burst = BurstSlc.burst_from_file("some_SAFE.zip")
```
I haven't worked how to eliminate code duplication between this `@classmethod` and `sentinel1_reader.burst_from_xml`.
2. Reduce the number of parameters needed to initialize a `Sentinel1BurstSlc` by grouping parameters .e.g. grouping [VRT params](https://github.com/opera-adt/s1-reader/blob/e4f52cefea65ad2cd1ca5441959738c3b5fba91d/src/sentinel1_reader/sentinel1_burst_slc.py#L187-L195) into a single `simplenamespce`. By doing so, accessing `tiff_path` goes from `burst.tiff_path` to `burst.vrt_params.tiff_path`.

I'll keep examining the code for a more concrete solution.
